File: lib/flash_profile/pattern.ex
defmodule FlashProfile.Pattern do
  @moduledoc """
  Implementation of patterns for FlashProfile.

  A **pattern** is simply a sequence of atoms. The pattern `Empty` denotes an empty
  sequence, which only matches the empty string `ε`. We use the concatenation
  operator `◇` for sequencing atoms.

  ## Pattern Matching Rules

  Pattern `P` describes string `s` iff:
  - `s ≠ ε` (non-empty) OR `s = ε` and `P` is empty
  - `∀i ∈ {1,...,k}: αi(si) > 0` (each atom matches a non-empty prefix)
  - `sk+1 = ε` (entire string consumed)

  Where `s1 = s` and `si+1 = si[αi(si):]` (remaining suffix after matching atom `αi`)

  ## Matching Algorithm

  Patterns match **greedily from left to right**:
  1. Start at position `0` of the string
  2. For each atom in order:
     - Call `atom.match(remaining_string)`
     - If result is `0`, pattern doesn't match
     - Otherwise, consume that many characters and continue
  3. After all atoms, check if entire string is consumed

  ## Examples

  ```elixir
  iex> alias FlashProfile.{Pattern, Atom}
  iex> alias FlashProfile.Atoms.CharClass
  iex> digit = CharClass.digit()
  iex> upper = CharClass.upper()
  iex> dash = Atom.constant("-")
  iex> pattern = [upper, dash, digit]
  iex> Pattern.matches?(pattern, "A-123")
  true
  iex> Pattern.matches?(pattern, "AB-123")
  true
  iex> Pattern.matches?(pattern, "A-")
  false
  ```
  """

  alias FlashProfile.Atom

  @type t :: [Atom.t()]

  @doc """
  Create an empty pattern (matches only empty string).

  ## Examples

      iex> FlashProfile.Pattern.empty()
      []
  """
  @spec empty() :: t()
  def empty(), do: []

  @doc """
  Check if a pattern matches a string entirely.

  Returns `true` if the pattern describes the string. An empty pattern `[]`
  only matches empty string `""`. Each atom must match a non-empty prefix,
  and together they must consume the entire string.

  ## Examples

  ```elixir
  iex> alias FlashProfile.{Pattern, Atom}
  iex> alias FlashProfile.Atoms.CharClass
  iex> digit = CharClass.digit()
  iex> pattern = [digit]
  iex> Pattern.matches?(pattern, "123")
  true
  iex> Pattern.matches?(pattern, "123abc")
  false
  iex> Pattern.matches?([], "")
  true
  iex> Pattern.matches?([], "abc")
  false
  ```
  """
  @spec matches?(t(), String.t()) :: boolean()
  def matches?(pattern, string) when is_list(pattern) and is_binary(string) do
    case do_match(pattern, string, 0, []) do
      {:ok, _matches} -> true
      {:error, :no_match} -> false
    end
  end

  @doc """
  Match a pattern against a string, returning match details.

  Returns `{:ok, matches}` with list of `{atom, matched_substring, start_pos, length}`
  or `{:error, :no_match}` if pattern doesn't match.

  ## Examples

  ```elixir
  iex> alias FlashProfile.{Pattern, Atom}
  iex> alias FlashProfile.Atoms.CharClass
  iex> digit = CharClass.digit()
  iex> upper = CharClass.upper()
  iex> dash = Atom.constant("-")
  iex> pattern = [upper, dash, digit]
  iex> {:ok, matches} = Pattern.match(pattern, "A-123")
  iex> length(matches)
  3
  iex> Pattern.match(pattern, "invalid")
  {:error, :no_match}
  ```
  """
  @spec match(t(), String.t()) ::
          {:ok, list({Atom.t(), String.t(), non_neg_integer(), non_neg_integer()})}
          | {:error, :no_match}
  def match(pattern, string) when is_list(pattern) and is_binary(string) do
    do_match(pattern, string, 0, [])
  end

  @doc """
  Get the lengths matched by each atom for a string.

  Returns list of lengths, or `nil` if pattern doesn't match.
  Used for cost calculation.

  ## Examples

  ```elixir
  iex> alias FlashProfile.{Pattern, Atom}
  iex> alias FlashProfile.Atoms.CharClass
  iex> digit = CharClass.digit()
  iex> upper = CharClass.upper()
  iex> pattern = [upper, digit]
  iex> Pattern.match_lengths(pattern, "A123")
  [1, 3]
  iex> Pattern.match_lengths(pattern, "invalid")
  nil
  ```
  """
  @spec match_lengths(t(), String.t()) :: [non_neg_integer()] | nil
  def match_lengths(pattern, string) when is_list(pattern) and is_binary(string) do
    case do_match(pattern, string, 0, []) do
      {:ok, matches} ->
        matches
        |> Enum.map(fn {_atom, _matched, _pos, length} -> length end)

      {:error, :no_match} ->
        nil
    end
  end

  @doc """
  Format a pattern as a human-readable string.

  ## Display Format

  - Constant strings: in quotes, e.g., `"PMC"`
  - Fixed-width char class: `Name×N`, e.g., `Digit×4`
  - Variable-width char class: `Name+`, e.g., `Lower+`
  - Atoms separated by `◇`

  ## Examples

  ```elixir
  iex> alias FlashProfile.{Pattern, Atom}
  iex> alias FlashProfile.Atoms.CharClass
  iex> digit = CharClass.digit()
  iex> upper = CharClass.upper()
  iex> dash = Atom.constant("-")
  iex> pattern = [upper, dash, digit]
  iex> Pattern.to_string(pattern)
  "Upper+ ◇ \\"-\\" ◇ Digit+"
  ```
  """
  @spec to_string(t()) :: String.t()
  def to_string(pattern) when is_list(pattern) do
    pattern
    |> Enum.map(&format_atom/1)
    |> Enum.join(" ◇ ")
  end

  @doc """
  Concatenate two patterns.

  ## Examples

      iex> alias FlashProfile.{Pattern, Atom}
      iex> alias FlashProfile.Atoms.CharClass
      iex> digit = CharClass.digit()
      iex> upper = CharClass.upper()
      iex> p1 = [upper]
      iex> p2 = [digit]
      iex> result = Pattern.concat(p1, p2)
      iex> length(result)
      2
  """
  @spec concat(t(), t()) :: t()
  def concat(pattern1, pattern2) when is_list(pattern1) and is_list(pattern2) do
    pattern1 ++ pattern2
  end

  @doc """
  Append an atom to a pattern.

  ## Examples

      iex> alias FlashProfile.{Pattern, Atom}
      iex> alias FlashProfile.Atoms.CharClass
      iex> digit = CharClass.digit()
      iex> upper = CharClass.upper()
      iex> pattern = [upper]
      iex> result = Pattern.append(pattern, digit)
      iex> length(result)
      2
  """
  @spec append(t(), Atom.t()) :: t()
  def append(pattern, atom) when is_list(pattern) do
    pattern ++ [atom]
  end

  @doc """
  Get the number of atoms in a pattern.

  ## Examples

      iex> alias FlashProfile.{Pattern, Atom}
      iex> alias FlashProfile.Atoms.CharClass
      iex> digit = CharClass.digit()
      iex> upper = CharClass.upper()
      iex> pattern = [upper, digit]
      iex> Pattern.length(pattern)
      2
      iex> Pattern.length([])
      0
  """
  @spec length(t()) :: non_neg_integer()
  def length(pattern) when is_list(pattern) do
    Kernel.length(pattern)
  end

  @doc """
  Check if pattern is empty.

  ## Examples

      iex> FlashProfile.Pattern.empty?([])
      true
      iex> alias FlashProfile.Atoms.CharClass
      iex> FlashProfile.Pattern.empty?([CharClass.digit()])
      false
  """
  @spec empty?(t()) :: boolean()
  def empty?(pattern) when is_list(pattern), do: pattern == []

  @doc """
  Get the first atom of a pattern.

  Returns nil if pattern is empty.

  ## Examples

      iex> alias FlashProfile.{Pattern, Atom}
      iex> alias FlashProfile.Atoms.CharClass
      iex> digit = CharClass.digit()
      iex> upper = CharClass.upper()
      iex> pattern = [upper, digit]
      iex> first = Pattern.first(pattern)
      iex> first.name
      "Upper"
      iex> Pattern.first([])
      nil
  """
  @spec first(t()) :: Atom.t() | nil
  def first([atom | _]), do: atom
  def first([]), do: nil

  @doc """
  Get the last atom of a pattern.

  Returns nil if pattern is empty.

  ## Examples

      iex> alias FlashProfile.{Pattern, Atom}
      iex> alias FlashProfile.Atoms.CharClass
      iex> digit = CharClass.digit()
      iex> upper = CharClass.upper()
      iex> pattern = [upper, digit]
      iex> last = Pattern.last(pattern)
      iex> last.name
      "Digit"
      iex> Pattern.last([])
      nil
  """
  @spec last(t()) :: Atom.t() | nil
  def last(pattern) when is_list(pattern) do
    List.last(pattern)
  end

  ## Private Functions

  # Match pattern against string, accumulating match details
  # Returns {:ok, matches} or {:error, :no_match}
  defp do_match([], "", _pos, acc) do
    {:ok, Enum.reverse(acc)}
  end

  defp do_match([], _remaining, _pos, _acc) do
    {:error, :no_match}
  end

  defp do_match([atom | rest], string, pos, acc) do
    case Atom.match(atom, string) do
      0 ->
        {:error, :no_match}

      len ->
        matched = String.slice(string, 0, len)
        remaining = String.slice(string, len..-1//1)
        do_match(rest, remaining, pos + len, [{atom, matched, pos, len} | acc])
    end
  end

  # Format an atom for display in a pattern string
  defp format_atom(%Atom{type: :constant, params: %{string: str}}) do
    # Constant strings shown in quotes
    inspect(str)
  end

  defp format_atom(%Atom{type: :char_class, name: name, params: %{width: width}})
       when width > 0 do
    # Fixed-width: Name×N
    "#{name}×#{width}"
  end

  defp format_atom(%Atom{type: :char_class, name: name, params: %{width: 0}}) do
    # Variable-width: Name+
    "#{name}+"
  end

  defp format_atom(%Atom{name: name}) do
    # Fallback: just use the name
    name
  end
end

File: native/flash_profile/pattern.zig
const std = @import("std");
const atom_mod = @import("atom.zig");
const Atom = atom_mod.Atom;
const types = @import("types.zig");

/// A pattern is a sequence of atoms.
///
/// From the FlashProfile paper (Definition 4.3):
/// "A pattern is simply a sequence of atoms. The pattern Empty denotes an empty
/// sequence, which only matches the empty string ε."
///
/// Pattern P describes string s iff:
/// - s ≠ ε (non-empty) OR s = ε and P is empty
/// - ∀i ∈ {1,...,k}: αi(si) > 0 (each atom matches a non-empty prefix)
/// - sk+1 = ε (entire string consumed)
///
/// Matching is greedy left-to-right:
/// 1. Start at position 0
/// 2. For each atom, match against remaining string
/// 3. If match fails (returns null), pattern doesn't match
/// 4. After all atoms, entire string must be consumed
pub const Pattern = struct {
    /// Sequence of atoms
    atoms: []const Atom,

    /// Create an empty pattern (matches only empty string).
    pub fn empty() Pattern {
        return .{ .atoms = &[_]Atom{} };
    }

    /// Create a pattern from a slice of atoms.
    pub fn init(atoms: []const Atom) Pattern {
        return .{ .atoms = atoms };
    }

    /// Check if this pattern matches a string entirely.
    ///
    /// Returns true if:
    /// - Empty pattern and empty string, OR
    /// - Each atom matches a non-empty prefix and entire string is consumed
    ///
    /// Examples:
    /// - Pattern [] matches ""
    /// - Pattern [Digit] matches "123" but not "123abc"
    pub fn matches(self: Pattern, string: []const u8) bool {
        return self.matchInternal(string) != null;
    }

    /// Match pattern and return the lengths matched by each atom.
    /// Returns null if pattern doesn't match.
    ///
    /// This is used for cost calculation - we need to know how much
    /// each atom matched to compute dynamic costs.
    pub fn matchLengths(self: Pattern, allocator: std.mem.Allocator, string: []const u8) !?[]usize {
        // Empty pattern only matches empty string
        if (self.atoms.len == 0) {
            return if (string.len == 0) try allocator.alloc(usize, 0) else null;
        }

        // Allocate array to hold lengths
        const lengths = try allocator.alloc(usize, self.atoms.len);
        errdefer allocator.free(lengths);

        var pos: usize = 0;

        for (self.atoms, 0..) |atom, i| {
            const remaining = string[pos..];
            const len = atom.match(remaining) orelse {
                allocator.free(lengths);
                return null;
            };

            lengths[i] = len;
            pos += len;
        }

        // Entire string must be consumed
        if (pos != string.len) {
            allocator.free(lengths);
            return null;
        }

        return lengths;
    }

    /// Internal matching that checks if pattern matches (for matches() function).
    /// Uses stack-allocated buffer, only returns bool-compatible result.
    fn matchInternal(self: Pattern, string: []const u8) ?[]const usize {
        // Empty pattern only matches empty string
        if (self.atoms.len == 0) {
            return if (string.len == 0) &[_]usize{} else null;
        }

        // Buffer to store match lengths (max 64 atoms per pattern)
        // Note: This is only used for the matches() function which just checks
        // if the return is non-null. The actual values are not used after return.
        var lengths_buffer: [64]usize = undefined;
        if (self.atoms.len > lengths_buffer.len) {
            // Pattern too long
            return null;
        }

        var pos: usize = 0;

        for (self.atoms, 0..) |atom, i| {
            const remaining = string[pos..];
            const len = atom.match(remaining) orelse return null;

            lengths_buffer[i] = len;
            pos += len;
        }

        // Entire string must be consumed
        if (pos != string.len) {
            return null;
        }

        // Return non-null to indicate match (values not used)
        return lengths_buffer[0..self.atoms.len];
    }
};

/// Helper function to check if pattern matches, with atom array.
/// Convenience wrapper for creating a Pattern on the fly.
pub fn matches(atoms: []const Atom, string: []const u8) bool {
    const pattern = Pattern.init(atoms);
    return pattern.matches(string);
}

/// Helper function to get match lengths for a pattern.
pub fn matchLengths(allocator: std.mem.Allocator, atoms: []const Atom, string: []const u8) !?[]usize {
    const pattern = Pattern.init(atoms);
    return pattern.matchLengths(allocator, string);
}

// ============================================================================
// Tests
// ============================================================================

test "Empty pattern" {
    const empty_pattern = Pattern.empty();

    try std.testing.expect(empty_pattern.matches(""));
    try std.testing.expect(!empty_pattern.matches("abc"));
}

test "Single atom pattern" {
    const d = atom_mod.digit();
    const pattern = Pattern.init(&[_]Atom{d});

    try std.testing.expect(pattern.matches("123"));
    try std.testing.expect(!pattern.matches("123abc"));
    try std.testing.expect(!pattern.matches("abc"));
}

test "Multi-atom pattern" {
    const u = atom_mod.upper();
    const dash = atom_mod.constant("Dash", "-");
    const d = atom_mod.digit();

    const pattern = Pattern.init(&[_]Atom{ u, dash, d });

    // Match: Upper+ "-" Digit+
    try std.testing.expect(pattern.matches("A-123"));
    try std.testing.expect(pattern.matches("ABC-456"));
    try std.testing.expect(!pattern.matches("A-"));
    try std.testing.expect(!pattern.matches("a-123"));
    try std.testing.expect(!pattern.matches("A123"));
}

test "Match lengths" {
    const u = atom_mod.upper();
    const d = atom_mod.digit();

    const pattern = Pattern.init(&[_]Atom{ u, d });

    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    // "A123" -> Upper matches 1, Digit matches 3
    const lengths = try pattern.matchLengths(allocator, "A123");
    try std.testing.expect(lengths != null);
    try std.testing.expectEqual(@as(usize, 2), lengths.?.len);
    try std.testing.expectEqual(@as(usize, 1), lengths.?[0]);
    try std.testing.expectEqual(@as(usize, 3), lengths.?[1]);

    // "ABC123" -> Upper matches 3, Digit matches 3
    const lengths2 = try pattern.matchLengths(allocator, "ABC123");
    try std.testing.expect(lengths2 != null);
    try std.testing.expectEqual(@as(usize, 2), lengths2.?.len);
    try std.testing.expectEqual(@as(usize, 3), lengths2.?[0]);
    try std.testing.expectEqual(@as(usize, 3), lengths2.?[1]);

    // "invalid" -> no match
    const no_match = try pattern.matchLengths(allocator, "invalid");
    try std.testing.expect(no_match == null);
}

test "Fixed-width atoms in pattern" {
    const u = atom_mod.upper();
    const d2 = atom_mod.withFixedWidth(atom_mod.digit(), 2);

    const pattern = Pattern.init(&[_]Atom{ u, d2 });

    // Upper+ Digit×2
    try std.testing.expect(pattern.matches("A12"));
    try std.testing.expect(pattern.matches("ABC12"));
    try std.testing.expect(!pattern.matches("A1"));
    try std.testing.expect(!pattern.matches("A123"));
}

test "Helper functions" {
    const d = atom_mod.digit();
    const atoms = [_]Atom{d};

    try std.testing.expect(matches(&atoms, "123"));
    try std.testing.expect(!matches(&atoms, "abc"));

    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const lengths = try matchLengths(allocator, &atoms, "456");
    try std.testing.expect(lengths != null);
    try std.testing.expectEqual(@as(usize, 1), lengths.?.len);
    try std.testing.expectEqual(@as(usize, 3), lengths.?[0]);
}
